app/actions/user.ts

```diff
@@ -1,171 +1,153 @@
-import { providers } from "ethers";
+import { providers, ethers } from "ethers";
 import { Thunk } from "state";
 
-import { addresses } from "@klimadao/lib/constants";
 import {
   formatUnits,
-  trimStringDecimals,
   getContract,
+  createContractsObject,
+  getAllowance,
+  getTokensFromSpender,
+  getTokenDecimals,
 } from "@klimadao/lib/utils";
-import {
-  setBalance,
-  setExerciseAllowance,
-  setStakeAllowance,
-  setWrapAllowance,
-  setDomains,
-} from "state/user";
+import { AllowancesFormatted } from "@klimadao/lib/types/allowances";
+import { setBalance, updateAllowances, setDomains } from "state/user";
 
 import { getKns, getEns } from "./utils";
 
+const assets = [
+  "bct",
+  "nct",
+  "mco2",
+  "ubo",
+  "nbo",
+  "usdc",
+  "klima",
+  "sklima",
+  "wsklima",
+  "pklima",
+] as const;
+type Asset = typeof assets[number];
+
+const spenders = [
+  "staking_helper",
+  "staking",
+  "wsklima",
+  "pklima_exercise",
+] as const;
+
+type TokenValueFormatted = {
+  [K in Asset]: string;
+};
+
+const getBalance = async (params: {
+  token: Asset;
+  contract: ethers.Contract;
+  address: string;
+}): Promise<TokenValueFormatted> => {
+  try {
+    const balance = await params.contract.balanceOf(params.address);
+    const decimals = getTokenDecimals(params.token);
+    return {
+      [params.token]: formatUnits(balance, decimals),
+    } as TokenValueFormatted;
+  } catch (e) {
+    console.error(`Error in getBalance for token: ${params.token}`);
+    return Promise.reject(e);
+  }
+};
+
 export const loadAccountDetails = (params: {
   provider: providers.JsonRpcProvider;
   address: string;
   onRPCError: () => void;
 }): Thunk => {
   return async (dispatch) => {
     try {
-      const bctContract = getContract({
-        contractName: "bct",
-        provider: params.provider,
-      });
-      const nctContract = getContract({
-        contractName: "nct",
-        provider: params.provider,
-      });
-      const mco2Contract = getContract({
-        contractName: "mco2",
-        provider: params.provider,
-      });
-      const uboContract = getContract({
-        contractName: "ubo",
-        provider: params.provider,
-      });
-      const nboContract = getContract({
-        contractName: "nbo",
-        provider: params.provider,
-      });
-      const usdcContract = getContract({
-        contractName: "usdc",
-        provider: params.provider,
-      });
-      const klimaContract = getContract({
-        contractName: "klima",
-        provider: params.provider,
-      });
-      const sklimaContract = getContract({
-        contractName: "sklima",
-        provider: params.provider,
-      });
-      const wsklimaContract = getContract({
-        contractName: "wsklima",
-        provider: params.provider,
-      });
-      const pKlimaContract = getContract({
-        contractName: "pklima",
+      const assetsContracts = createContractsObject({
+        contracts: assets as any,
```

Comment:
This is risky, what is the type error saying?
it looks like the `assets` array is strongly typed, so perhaps there is something missing in `contracts` type?

Reply:
It's refactored